cpp/src/communicator/ucxx.cpp

```diff
@@ -1091,28 +1091,27 @@ std::unique_ptr<Communicator::Future> UCXX::send(
 std::unique_ptr<Communicator::Future> UCXX::send(
     std::unique_ptr<Buffer> msg, Rank rank, Tag tag
 ) {
-    if (!msg->is_ready()) {
-        logger().warn("msg is not ready. This is irrecoverable, terminating.");
-        std::terminate();
-    }
+    RAPIDSMPF_EXPECTS(msg->is_latest_write_done(), "msg must be ready");
+    auto data = const_cast<std::byte*>(msg->data());
     auto req = get_endpoint(rank)->tagSend(
-        msg->data(), msg->size, tag_with_rank(shared_resources_->rank(), tag)
+        data, msg->size, tag_with_rank(shared_resources_->rank(), tag)
     );
     return std::make_unique<Future>(req, std::move(msg));
 }
 
 std::unique_ptr<Communicator::Future> UCXX::recv(
     Rank rank, Tag tag, std::unique_ptr<Buffer> recv_buffer
 ) {
-    if (!recv_buffer->is_ready()) {
-        logger().warn("recv_buffer is not ready. This is irrecoverable, terminating.");
-        std::terminate();
-    }
-    auto req = get_endpoint(rank)->tagRecv(
-        recv_buffer->data(),
-        recv_buffer->size,
-        tag_with_rank(rank, tag),
-        ::ucxx::TagMaskFull
+    RAPIDSMPF_EXPECTS(recv_buffer->is_latest_write_done(), "msg must be ready");
+    auto req = recv_buffer->write_access(
+        recv_buffer->stream(), [&](std::byte* recv_buffer_data) {
+            return get_endpoint(rank)->tagRecv(
+                recv_buffer_data,
+                recv_buffer->size,
+                tag_with_rank(rank, tag),
+                ::ucxx::TagMaskFull
+            );
+        }
```

Comment:
*issue*: This has the same kind of issue as the same thing in `MPI::recv`.

Also, this also is indicative that `is_latest_write_done` is a bit of a misnomer. In this case what it means is "is the allocation done" (and I suppose, if there are writes to this buffer, are they flushed?)